File: app/src/main/java/dev/ragnarok/fenrir/crypt/ExchangeMessage.java
package dev.ragnarok.fenrir.crypt;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;

import dev.ragnarok.fenrir.util.ParcelUtils;

public class ExchangeMessage implements Parcelable {

    public static final Creator<ExchangeMessage> CREATOR = new Creator<ExchangeMessage>() {
        @Override
        public ExchangeMessage createFromParcel(Parcel in) {
            return new ExchangeMessage(in);
        }

        @Override
        public ExchangeMessage[] newArray(int size) {
            return new ExchangeMessage[size];
        }
    };
    @SerializedName("v")
    private int version;
    @SerializedName("sid")
    private long sessionId;
    @SerializedName("public_key")
    private String publicKey;
    @SerializedName("aes_key")
    private String aesKey;
    @SessionState
    @SerializedName("session_state")
    private int senderSessionState;
    @SerializedName("error_code")
    private int errorCode;
    @KeyLocationPolicy
    @SerializedName("klp")
    private Integer keyLocationPolicy;

    @SuppressWarnings("unused")
    public ExchangeMessage() {
    }

    private ExchangeMessage(Builder builder) {
        publicKey = builder.publicKey;
        version = builder.version;
        sessionId = builder.sessionId;
        aesKey = builder.aesKey;
        senderSessionState = builder.sessionState;
        errorCode = builder.errorCode;
        keyLocationPolicy = builder.keyLocationPolicy;
    }

    protected ExchangeMessage(Parcel in) {
        version = in.readInt();
        sessionId = in.readLong();
        publicKey = in.readString();
        aesKey = in.readString();

        @SessionState
        int s = in.readInt();
        senderSessionState = s;
        errorCode = in.readInt();

        @KeyLocationPolicy
        Integer klp = ParcelUtils.readObjectInteger(in);
        keyLocationPolicy = klp;
    }

    public boolean isError() {
        return errorCode != 0;
    }

    @NonNull
    @Override
    public String toString() {
        return "RSA" + new Gson().toJson(this);
    }

    public int getVersion() {
        return version;
    }

    public String getPublicKey() {
        return publicKey;
    }

    public long getSessionId() {
        return sessionId;
    }

    public String getAesKey() {
        return aesKey;
    }

    @SessionState
    public int getSenderSessionState() {
        return senderSessionState;
    }

    @KeyLocationPolicy
    public int getKeyLocationPolicy() {
        return keyLocationPolicy;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(version);
        dest.writeLong(sessionId);
        dest.writeString(publicKey);
        dest.writeString(aesKey);
        dest.writeInt(senderSessionState);
        dest.writeInt(errorCode);
        ParcelUtils.writeObjectInteger(dest, keyLocationPolicy);
    }

    public int getErrorCode() {
        return errorCode;
    }

    public static final class Builder {

        private final int version;
        private final long sessionId;
        @SessionState
        private final int sessionState;
        private String publicKey;
        private String aesKey;
        @KeyLocationPolicy
        private Integer keyLocationPolicy;
        private int errorCode;

        public Builder(int version, long sessionId, @SessionState int senderSessionState) {
            this.version = version;
            this.sessionId = sessionId;
            sessionState = senderSessionState;
        }

        public Builder setPublicKey(String publicKey) {
            this.publicKey = publicKey;
            return this;
        }

        public Builder setAesKey(String aesKey) {
            this.aesKey = aesKey;
            return this;
        }

        public Builder setErrorCode(int errorCode) {
            this.errorCode = errorCode;
            return this;
        }

        public ExchangeMessage create() {
            return new ExchangeMessage(this);
        }

        public Builder setKeyLocationPolicy(@KeyLocationPolicy int keyLocationPolicy) {
            this.keyLocationPolicy = keyLocationPolicy;
            return this;
        }
    }

}

File: app/src/main/kotlin/dev/ragnarok/fenrir/crypt/ExchangeMessage.kt
package dev.ragnarok.fenrir.crypt

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import dev.ragnarok.fenrir.orZero
import dev.ragnarok.fenrir.util.ParcelUtils.readObjectInteger
import dev.ragnarok.fenrir.util.ParcelUtils.writeObjectInteger

class ExchangeMessage : Parcelable {
    @SerializedName("v")
    var version = 0
        private set

    @SerializedName("sid")
    var sessionId: Long = 0
        private set

    @SerializedName("public_key")
    var publicKey: String? = null
        private set

    @SerializedName("aes_key")
    var aesKey: String? = null
        private set

    @get:SessionState
    @SessionState
    @SerializedName("session_state")
    var senderSessionState = 0
        private set

    @SerializedName("error_code")
    var errorCode = 0
        private set

    @KeyLocationPolicy
    @SerializedName("klp")
    private var keyLocationPolicy: Int? = null

    @Suppress("UNUSED")
    constructor()
    private constructor(builder: Builder) {
        publicKey = builder.publicKey
        version = builder.version
        sessionId = builder.sessionId
        aesKey = builder.aesKey
        senderSessionState = builder.sessionState
        errorCode = builder.errorCode
        keyLocationPolicy = builder.keyLocationPolicy
    }

    private constructor(`in`: Parcel) {
        version = `in`.readInt()
        sessionId = `in`.readLong()
        publicKey = `in`.readString()
        aesKey = `in`.readString()
        @SessionState val s = `in`.readInt()
        senderSessionState = s
        errorCode = `in`.readInt()
        @KeyLocationPolicy val klp = readObjectInteger(`in`)
        keyLocationPolicy = klp
    }

    val isError: Boolean
        get() = errorCode != 0

    override fun toString(): String {
        return "RSA" + Gson().toJson(this)
    }

    @KeyLocationPolicy
    fun getKeyLocationPolicy(): Int {
        return keyLocationPolicy.orZero()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(version)
        dest.writeLong(sessionId)
        dest.writeString(publicKey)
        dest.writeString(aesKey)
        dest.writeInt(senderSessionState)
        dest.writeInt(errorCode)
        writeObjectInteger(dest, keyLocationPolicy)
    }

    class Builder(
        val version: Int,
        val sessionId: Long,
        @SessionState val sessionState: Int
    ) {
        var publicKey: String? = null
        var aesKey: String? = null

        @KeyLocationPolicy
        var keyLocationPolicy: Int? = null
        var errorCode = 0
        fun setPublicKey(publicKey: String?): Builder {
            this.publicKey = publicKey
            return this
        }

        fun setAesKey(aesKey: String?): Builder {
            this.aesKey = aesKey
            return this
        }

        fun setErrorCode(errorCode: Int): Builder {
            this.errorCode = errorCode
            return this
        }

        fun create(): ExchangeMessage {
            return ExchangeMessage(this)
        }

        fun setKeyLocationPolicy(@KeyLocationPolicy keyLocationPolicy: Int): Builder {
            this.keyLocationPolicy = keyLocationPolicy
            return this
        }
    }

    companion object CREATOR : Parcelable.Creator<ExchangeMessage> {
        override fun createFromParcel(parcel: Parcel): ExchangeMessage {
            return ExchangeMessage(parcel)
        }

        override fun newArray(size: Int): Array<ExchangeMessage?> {
            return arrayOfNulls(size)
        }
    }
}